packages/client/src/client.spec.js

```diff
@@ -3091,3 +3092,46 @@ describe('test_whitelabel_links__link_id__subuser_post', () => {
     return testRequest(request, 200);
   });
 });
+
+describe('setDataResidency', () => {
+  const testClient = require('./client');
+  let consoleWarnSpy;
+
+  beforeEach(() => {
+    consoleWarnSpy = sinon.spy(console, 'warn');
+  });
+  afterEach(() => {
+    console.warn.restore();
+  });
+
+  it('should send to host EU', () => {
+    testClient.setDataResidency('eu');
+    expect(testClient.defaultRequest.baseUrl).to.equal('https://api.eu.sendgrid.com/');
+  });
+  it('should send to host Global/default', () => {
+    testClient.setDataResidency('global');
+    expect(testClient.defaultRequest.baseUrl).to.equal('https://api.sendgrid.com/');
+  });
+  it('should override the existing set hostname, if data residency setter is called after', () => {
+    testClient.setApiKey('SG.1234567890');
+    testClient.setDataResidency('eu');
+    expect(testClient.defaultRequest.baseUrl).to.equal('https://api.eu.sendgrid.com/');
+  });
+  it('should give a warning if the provided value is not allowed', () => {
+    testClient.setDataResidency('');
+    expect(consoleWarnSpy.calledOnce).to.equal(true);
+  });
+  it('should give a warning if the provided value is null', () => {
+    testClient.setDataResidency(null);
+    expect(consoleWarnSpy.calledOnce).to.equal(true);
+  });
+  it('should give precedence to the order of execution', () => {
+    testClient.setDataResidency('eu');
+    testClient.setApiKey('SG.1234567890');
```

Comment:
why would `setApiKey` override the `hostname`? If it really needs to, is this clearly documented somewhere? But I personally think this should not change the `hostname`.

Reply:
For some reason, setApiKey() also [sets the baseUrl](https://github.com/sendgrid/sendgrid-nodejs/blob/b8125d86134157a462c2e24816cfa56677ab76a8/packages/client/src/classes/client.js#L41) while in action. I do not know why exactly was this done, but it looks like for each auth call for [Sendgrid](https://github.com/sendgrid/sendgrid-nodejs/blob/b8125d86134157a462c2e24816cfa56677ab76a8/packages/client/src/classes/client.js#L39) or [Twilio](https://github.com/sendgrid/sendgrid-nodejs/blob/b8125d86134157a462c2e24816cfa56677ab76a8/packages/client/src/classes/client.js#L48), it was needed.

Reply:
I can add this in the **Limitations** section in usage doc.

Reply:
Done ^

Reply:
I see. I think that was done to automatically switch between the twilio vs sendgrid endpoint based on the type of auth used. It makes sense for that specific use case but I think we should try to elegantly handle this on the regional sendgrid side. How does it work for other libraries?

I'm thinking something along the lines of:
- set to global endpoint if `baseUrl` is not yet set to sendgrid
- leave the value as is if it's already using one of the sendgrid endpoints (eu or global).

OR we can track the region that was set and `setApiKey` will set the `baseUrl` using the correct region.

cc: @tbischel in case you have any input

Reply:
Other libraries don't have this situation, they have different classes for Twilio and Sendgrid, inheriting the same parent class.

Reply:
I agree with Gladys, this becomes a bad user experience as is when any combination of setApiKey and setting the data residency happen in different orders.  I'd be inclined to track the region, where setApiKey uses the correct region, and the default region is global.